File: swift/swift/resource/memory-management.swift
//
//  memory-management.swift
//  swift
//
//  Created by ke on 3/24/24.
//

import Foundation

//1.using ARC
class ARC {
    
}

func usingARC() {
    var ref1: ARC? = ARC()
    var ref2: ARC? = ref1

    ref1 = nil
    ref2 = nil
}

//2.reference cycles problem
class A {
    var b: B?
}

class B {
    var a: A?
}

func rc(){
    var ca: A? = A()
    var cb: B? = B()
    ca!.b = cb
    cb!.a = ca
    
    ca = nil
    cb = nil
}

File: go/resource/memory-management.go
package resource

import "fmt"

// 1.using GC
type Automate struct {
}

func usingGC() {
	a := Automate{}
	fmt.Println(a)
}

//2.reference cycles problem,In this specific case,
/*when you set both ca and cb to nil,
you're effectively breaking the cyclic reference between A and B.
Since there are no other references to the memory allocated for A and B,
the garbage collector will eventually detect that these objects
are unreachable and free their memory.*/
type A struct {
	b *B
}

type B struct {
	a *A
}

func rc() {
	ca := &A{}
	cb := &B{}

	ca.b = cb
	cb.a = ca

	ca = nil
	cb = nil
}
